src/radical/pilot/agent/executing/base.py

```diff
@@ -175,5 +188,55 @@ def handle_timeout(self, task):
                 self._to_tasks.append([to, time.time(), task])
 
 
+    # --------------------------------------------------------------------------
+    #
+    def advance_tasks(self, tasks, state, publish, push, ts=None):
+        '''
+        sort tasks into different buckets, depending on their origin.
+        That origin will determine where tasks which completed execution
+        and end up here will be routed to:
+
+          - client: state update to update worker
+          - raptor: state update to `STATE_PUBSUB`
+          - agent : state update to `STATE_PUBSUB`
+
+        a fallback is not in place to enforce the specification of the
+        `origin` attributes for tasks.
+        '''
+
+
+        buckets = {'client': list(),
+                   'raptor': list(),
+                   'agent' : list()}
+
+        for task in ru.as_list(tasks):
+            buckets[task['origin']].append(task)
+
+        # we want any task which has a `raptor_id` set to show up in raptor's
+        # result callbacks
+        if state != rps.AGENT_EXECUTING:
+            for task in ru.as_list(tasks):
+                if task['description'].get('raptor_id'):
+                    if task not in buckets['raptor']:
+                        buckets['raptor'].append(task)
+
+        if buckets['client']:
+            self.advance(buckets['client'], state=state,
+                                            publish=publish, push=push, ts=ts)
+
+        if buckets['raptor']:
+            self.advance(buckets['client'], state=state,
+                                            publish=publish, push=False, ts=ts)
+            self.publish(rpc.STATE_PUBSUB, {'cmd': 'raptor_state_update',
+                                            'arg': buckets['raptor']})
+
+        if buckets['agent']:
+            self.advance(buckets['client'], state=state,
+                                            publish=publish, push=False, ts=ts)
+            self.publish(rpc.STATE_PUBSUB, {'cmd': 'agent_state_update',
```

Comment:
should we use this command in `_service_state_cb` instead of `update`?

p.s. do we need such extra command? `update` is already in, whenever `cb` check for tasks it can find `origin` on its own. Kinda duplication (?)

Reply:
Good catch - that was an intermediate state of the code where that was used, is indeed not needed anymore.  I removed it.